lisa/sut_orchestrator/baremetal/build.py

```diff
@@ -79,23 +104,68 @@ def copy(self, sources_path: List[Path], files_map: List[FileSchema]) -> None:
 
-    def _find_matched_files(
-        self, sources_path: List[Path], files_map: List[FileSchema]
-    ) -> Dict[str, FileSchema]:
-        all_files = []
-        match_files: Dict[str, FileSchema] = {}
-        for source_path in sources_path:
-            for root, _, files in os.walk(source_path):
-                for file in files:
-                    all_files.append(os.path.join(root, file))
 
-            for file_map in files_map:
-                file_path = rf"{source_path}\{file_map.source}".replace("\\", "\\\\")
-                pattern = re.compile(
-                    file_path,
-                    re.I | re.M,
+class TftpBuild(Build):
+    def __init__(self, runbook: TftpBuildSchema) -> None:
+        super().__init__(runbook)
+        self.pxe_runbook: TftpBuildSchema = self.runbook
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "tftp"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return TftpBuildSchema
+
+    def copy(self, sources_path: List[Path], files_map: List[FileSchema]) -> None:
+        assert self.pxe_runbook.connection, "The build server is not specified"
+
+        server_address = self.pxe_runbook.connection.address
+        server_port = self.pxe_runbook.connection.port
+        server_username = self.pxe_runbook.connection.username
+        server_password = self.pxe_runbook.connection.password
+
+        try:
+            transport = paramiko.Transport((server_address, server_port))
+            transport.connect(
+                username=server_username,
+                password=server_password,
+            )
+        except Exception as ex:
+            self._log.exception(
+                "failed to connect to {server_address}:{server_port}",
+                exc_info=ex,
+            )
+        sftp = paramiko.SFTPClient.from_transport(transport)
+        assert sftp, "sft client is not created"
+
+        for file, file_map in Build.find_matched_files(
+            sources_path,
+            files_map,
+        ).items():
+            if file_map.destination:
+                attrs = sftp.lstat(file_map.destination)
+                if attrs.st_mode and stat.S_ISDIR(attrs.st_mode):
+                    file_name = file_map.destination + "/" + file.rsplit("\\")[-1]
+                else:
+                    file_name = file_map.destination
+            else:
+                file_name = file.rsplit("\\")[-1]
+
+            try:
+                sftp.put(file, file_name)
+            except Exception as ex:
+                self._log.exception(
+                    "failed to copy file {file} to {file_name}",
+                    exc_info=ex,
                 )
-                for file in all_files:
-                    if pattern.match(file):
-                        match_files[file] = file_map
-        return match_files
+
+        try:
+            sftp.close()
```

Comment:
As above comment, it should use remote_copy tool. But the close pattern should like below to avoid leak resource.
```python
try:
      try:
            transport = paramiko.Transport((server_address, server_port))
            transport.connect(
                username=server_username,
                password=server_password,
            )
        except Exception as ex:
            self._log.exception(
                "failed to connect to {server_address}:{server_port}",
                exc_info=ex,
            )
    ...
final:
    try:
        if sftp:
            sftp.close()
        if transport:
            transport.close()
    except Exception as ex:
        self._log.exception(
            "failed to close the connection",
            exc_info=ex,
        )
```

Reply:
The change is done, please review it again, thanks